File: 1743.cpp
#include <iostream>
#include <string>
#include <map>
#include <algorithm>
#include <cmath>

int main(){
    int n;
    std::string def;
    std::cin >> def;
    n = def.length();

    std::map<char, int> count {};

    for (auto c : def){
        if (count.find(c) != count.end()){
            count[c] += 1;
        } else {
            count[c] = 1;
        }
    }

    std::string res {};

    for (int i = 0; i < n; i++){
        bool found = false;

        for (auto & c : count){
            if (c.second == 0) continue;
            if (res.length() != 0 && res.back() == c.first) continue;
            c.second -= 1;
            bool is_valid = true;
            for (auto & d : count){
                if (d.second > std::ceil((n - i - 1) / 2.F)){
                    is_valid = false;
                    break;
                }
            }
            if (is_valid){
                res += c.first;
                found = true;
                break;
            } else {
                c.second += 1;
            }
        }

        if (!found){
            std::cout << -1 << std::endl;
            return 0;
        }
    }

    std::cout << res << std::endl;
    return 0;
}

File: 1743.py
s = input()
n = len(s)
count = [0 for _ in range(26)]
res = ["" for _ in range(n)]
c = -1

for ch in s: count[ord(ch) - ord('A')] += 1

if not all(2 * v <= (n + 1) for v in count):
    print(-1)
    exit()

for i in range(n):
    found = False
    for j in range(26):
        char = chr(ord('A') + j)
        if count[j] == 0:
            continue
        if c != -1 and res[c] == char:
            continue
        count[j] -= 1
        is_valid = 2 * max(count) <= n - i
        if is_valid:
            res[i] = char
            found = True
            c += 1
            break
        count[j] += 1
    if not found:
        print(-1)
        exit()

print("".join(res))

